text/0024-track-validator-set-when-synchronizing.org

```diff
@@ -0,0 +1,142 @@
+       \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+ |--   \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+       \/   .~~~~~~~.     ,--+----.
+       __   : A B D :~~~~/ Era 3 /<--- Update validator set
+       \/   `~~~~~~~'   '----+--'
+       __                    |
+       \/                   [#]<------ Check A/B/D sigs
+       __                    :
+       \/                    :
+
+
+  +--------+-------------------------------------+
+  | Legend |                                     |
+  +--------+                                     |
+  |                                              |
+  |                          |                   |
+  |  .~~~~.                 ,+--.                |
+  |  :    : Validator Set  /   /   Switch Block  |
+  |  `~~~~'               '--+'                  |
+  |                          |                   |
+  |                                              |
+  |                          |                   |
+  |                         [#]    Normal Block  |
+  |                          |                   |
+  |                                              |
+  +----------------------------------------------+
+
+
+#+END_SRC
+
+#+RESULTS:
+[[file:images/0024/joining.svg]]
+
+As we can see above, a switch block can change the validator
+set. Joining nodes needs to track this.
+
+** Reference-level explanation
+:PROPERTIES:
+:CUSTOM_ID: reference-level-explanation
+:END:
+
+/Synchronization mode/ is mediated by the variant
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L85-L93][=linear_chain_sync::State::SyncingDescendants]]=.
+
+We could add a field =validators= to this with type
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/types/src/auction/types.rs#L15][=types::ValidatorWeights=]] (i.e. =BTreeMap<PublicKey, U512>=). 
+
+This would be updated as switch blocks are processed. To support this,
+A field =next_era_validators= with type =BTreeMap<VID,U512>= will be added
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/consensus/consensus_protocol.rs#L34-L42][=consensus_protocol::EraEnd=]].
+
+** Unresolved questions
+:PROPERTIES:
+:CUSTOM_ID: unresolved-questions
+:END:
+
+- While validators are now tracked, how will
+  blocks be checked as discussed in the [[#guide-level-explanation][Guide-level
+  explanation]]? This is addressed in [[https://github.com/casperlabs/ceps/pull/1234][casperlabs/ceps#1234]].
+
+- What changes need to be made to transitioning from the
+  =SyncingDescendants= state to ordinary validation?  This is addressed
+  in [[https://github.com/casperlabs/ceps/pull/9999][casperlabs/ceps#9999]].
+ 
+- How will we enforce the new invariants?  What is the testing strategy?
+
+- What else is needed to support fast sync?
```

Comment:
Is this a rhetorical question? We know that we need at least the ability to download a global state at specific block.

Reply:
Ah, I see. It's only part of the whole design…